freqtrade/exchange/__init__.py

```diff
@@ -105,136 +94,227 @@ def validate_pairs(pairs: List[str]) -> None:
+        )
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not place buy order due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
 
-def sell(pair: str, rate: float, amount: float) -> str:
+def sell(pair: str, rate: float, amount: float) -> Dict:
     if _CONF['dry_run']:
         global _DRY_RUN_OPEN_ORDERS
         order_id = 'dry_run_sell_{}'.format(randint(0, 10**6))
         _DRY_RUN_OPEN_ORDERS[order_id] = {
             'pair': pair,
-            'rate': rate,
+            'price': rate,
             'amount': amount,
-            'type': 'LIMIT_SELL',
+            'type': 'limit',
+            'side': 'sell',
             'remaining': 0.0,
-            'opened': arrow.utcnow().datetime,
-            'closed': arrow.utcnow().datetime,
+            'datetime': arrow.utcnow().isoformat(),
+            'status': 'closed'
         }
-        return order_id
+        return {'id': order_id}
 
-    return _API.sell(pair, rate, amount)
+    try:
+        return _API.create_limit_sell_order(pair, amount, rate)
+    except ccxt.InsufficientFunds as e:
+        raise DependencyException(
+            'Insufficient funds to create limit sell order on market {}.'
+            'Tried to sell amount {} at rate {} (total {}).'
+            'Message: {}'.format(pair, amount, rate, rate*amount, e)
+        )
+    except ccxt.InvalidOrder as e:
+        raise DependencyException(
+            'Could not create limit sell order on market {}.'
+            'Tried to sell amount {} at rate {} (total {}).'
+            'Message: {}'.format(pair, amount, rate, rate*amount, e)
+        )
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not place sell order due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
 
 def get_balance(currency: str) -> float:
     if _CONF['dry_run']:
         return 999.9
 
-    return _API.fetch_balance()[currency]
+    # ccxt exception is already handled by get_balances
+    balances = get_balances()
+    return balances[currency]['free']
 
 
-def get_balances():
+def get_balances() -> dict:
     if _CONF['dry_run']:
-        return []
+        return {}
+
+    try:
+        balances = _API.fetch_balance()
+        # Remove additional info from ccxt results
+        balances.pop("info", None)
+        balances.pop("free", None)
+        balances.pop("total", None)
+        balances.pop("used", None)
+
+        return balances
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not get balance due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
-    return _API.fetch_balance()
 
-# @cached(TTLCache(maxsize=100, ttl=30))
```

Comment:
I'm not sure if it is a good idea to drop this decorator completely, because it saves a lot of redundant HTTP calls, maybe we should lower the `ttl` value

Reply:
It might be good to keep it but I don't think it should be a problem without it. `get_ticker()` is used 3 times in freqtradebot.py, once for buying and twice for selling. One of the usages while selling can use the previous ticker result found during the sell without caching. You don't want it to be cached anyway when checking bid prices during operation. Then it is used for various manual RPC calls, which are not called very often.

`get_ticker_history()` is only used in `get_signal()` during normal operation, and then it needs to return fresh data. 

Maybe I've missed something, but I think the cache should be cleared each "tick" of the bot anyway, so we have fresh data for checking buy/sell signal. So the only gain by caching is if a function is called multiple times per "tick", and that doesn't seem to be the case right now.

Reply:
The cached version should be used only from a call made in RPC, to avoid fetchning new bid/ask price when displaying the prrformace of the currently opened trades. It should not be used elsewhere.

Reply:
Yeah. The problem with cachetools (TTLCache and others) is that you can't remove a single item from the cache or force it to ignore the cache AFAIK. So for each pair in the whitelist we would have to clear the whole cache before fetching its ticker when analyzing buy/sell signals. After all pairs have been analyzed only the final pair would actually be cached for use in RPC.

So either we don't cache it, find some other cache library, or revert https://github.com/gcarq/freqtrade/pull/585/commits/954e5c095feb51a1f8535ed555b5db3551bad640

Reply:
@enenn Without caching chances are high that we hit an exchange rate limit (I don't know if cctx handles this under the hood), but we had this issue before and thats the reason we decided to cache some calls. The only mechanism left to avoid hitting the rate limit would be `process_throttle_secs`, but this doesn't help if you have a lot of pairs in the whitelist.  I think it is fine to cache the results for `lowest_possible_ticker_interval / 2 = 30 seconds` ( we could possibly lower this value to `lowest_possible_ticker_interval / 4`)

This implementation of TTLCache has its own ttl for each return value, it doesn't invalidate the whole cache buffer if one item is no longer valid.

> This class associates a time-to-live value with each item. Items that expire because they have exceeded their time-to-live will be removed automatically. If no expired items are there to remove, the least recently used items will be discarded first to make space when necessary. Trying to access an expired item will raise a KeyError.

http://cachetools.readthedocs.io/en/latest/#cachetools.TTLCache

Reply:
@gcarq I understand that each item has a TTL, but there is no way to know if you are getting a cached version or not with TTLcache, and there is no way to ensure a fresh value. Unless we are OK with getting a possible 30/15 second old ticker during the buy/sell signal stage? The old caching from 954e5c0 at least has a way to force a fresh value with the refresh argument.

 >I think it is fine to cache the results for lowest_possible_ticker_interval / 2 = 30 seconds ( we could possibly lower this value to lowest_possible_ticker_interval / 4)

I'd rather have fresh values for the buy/sell signal checking stage, and simply run the loop less often with `process_throttle_secs`. This way you are at least not acting on old market data, but rather acting less often. Perhaps the throttling could be weighted with the amount of pairs or something similar?

However, at this point this issue is getting somewhat complicated and I'm inclined to just revert 954e5c0 (which I guess worked for the old bot since python-bittrex had internal caching similar to what I implemented) and leave this caching issue to a different PR.

Reply:
> I don't know if cctx handles this under the hood

@gcarq ccxt does have internal rate-limiting. You need to set it `True` it when initializing the given exchange (which we don't do currently).

I leave this as a not to myself https://github.com/ccxt/ccxt/issues/1399#issuecomment-359233015

Reply:
@shusso Cool, thanks for the link. I'm fine with removing the decorator and let cctx handle the rate limits in the future